### ISSUE TITLE
Room for upload speed optimization ?

### ISSUE THREAD

**Issue by vadmeste** [minio/minio#2208]
| File size | 1 | 2 | 3 | 4 | 5 |
| --- | --- | --- | --- | --- | --- |
| 128mb.bin | 0m3.731s | 0m4.346s | 0m3.155s | 0m3.638s | 0m3.558s |
| 256mb.bin | 0m7.930s | 0m9.212s | 0m6.616s | 0m9.320s | 0m7.679s |
| 512mb.bin | 0m15.391s | 0m16.799s | 0m13.966s | 0m17.159s | 0m14.050s |
| 1gb.bin | 0m25.822s | 0m31.174s | 0m31.124s | 0m28.905s | 0m30.593s |
| 2gb.bin | 0m56.155s | 1m1.047s | 0m59.210s | 0m58.007s | 0m56.220s |
| 5gb.bin | 2m35.795s | 2m38.036s | 2m34.370s | 2m34.495s | 2m32.802s |
| 10gb.bin | 6m36.703s | 6m35.701s | 7m4.791s | 6m40.731s | 6m43.011s |
| 15gb.bin | 11m15.861s | 11m16.857s | 11m20.231s | 11m15.996s | 11m43.582s |
| 30gb.bin | 30m21.220s | 30m1.867s | 29m51.106s | 30m55.993s | 29m53.815s |

Tested on minio2.

Analyzing stats for 5gb, 10gb, 15gb, 30gb makes me wonder why time duration for uploading objects is not proportional to the size of the object.

Tests and stats generation are done by a shell script and can be redone as many times as needed

**Comment by vadmeste** [comment]
CPU profiling shows that main.readXLMeta() takes 14.6 sec when uploading a 2.5gb file but also takes 55.77 sec to upload 5gb file which not proportional. I wonder if we can make xl.json split by parts too so object file size won't matter.

**Comment by vadmeste** [comment]
```
(pprof) list main.readXLMeta
Total: 3.33mins
ROUTINE ======================== main.readXLMeta in /home/minio/gopath/src/github.com/minio/minio/xl-v1-utils.go
      20ms     55.77s (flat, cum) 27.87% of Total
         .          .     87:   }
         .          .     88:   return nums
         .          .     89:}
         .          .     90:
         .          .     91:// readXLMeta reads `xl.json` and returns back XL metadata structure.
         .       10ms     92:func readXLMeta(disk StorageAPI, bucket string, object string) (xlMeta xlMetaV1, err error) {
         .          .     93:   // Reads entire `xl.json`.
      10ms      770ms     94:   buf, err := disk.ReadAll(bucket, path.Join(object, xlMetaJSONFile))
         .          .     95:   if err != nil {
         .          .     96:           return xlMetaV1{}, err
         .          .     97:   }
         .          .     98:
         .          .     99:   // Unmarshal xl metadata.
      10ms     54.99s    100:   if err = json.Unmarshal(buf, &xlMeta); err != nil {
         .          .    101:           return xlMetaV1{}, err
         .          .    102:   }
         .          .    103:
         .          .    104:   // Return structured `xl.json`.
         .          .    105:   return xlMeta, nil
```

Time mostly spent in json.Unmarshal

**Comment by krishnasrinivas** [comment]
cool, looks like it's a known problem, https://github.com/pquerna/ffjson

**Comment by vadmeste** [comment]
@krishnasrinivas, the problem is not only with how fast the golang json parser.. but xl.json gets bigger as we upload new parts so naturally parsing json will slow down

**Comment by harshavardhana** [comment]
https://go-review.googlesource.com/#/c/24466/ , https://github.com/golang/go/issues/5683

**Comment by hackintoshrao** [comment]
Needs delicate changes with constant profiling  [5th August].

**Comment by hackintoshrao** [comment]
evaluating https://github.com/tidwall/gjson .

**Comment by hackintoshrao** [comment]
The benchmark [here](https://gist.github.com/hackintoshrao/a943af896f2fbb6ab97bd1acb98f7d60) compares the performance of `gjson` with the json.Unmarshal. gjson clearly has an edge by speed enchancement ranging from 20-100 times.

**Comment by hackintoshrao** [comment]
Even reconstructing the entire XLMetaV1{} using `gjson` is faster than one time `unmarshal`, Here is the benchmark to prove it https://gist.github.com/hackintoshrao/20bb29c8d290c32c7009dcb3a4600e48. Will draft a first version of change with `readAllXLMeta` returning `xlMetaV1{}` constructed using `gjson`, @vadmeste  can run the same test and see if there's any performance gain.

**Comment by harshavardhana** [comment]
Can you paste the benchmarks on how faster ?  I mean the output.

**Comment by hackintoshrao** [comment]
BenchmarkParseAllGJSON\* - Benchamark for constructing XLV1Meta{} from `gjson`. 
BenchmarkParseAllUnMarshal\* - Benchmark with fields obtained from  `json.unmarshal`.

```
BenchmarkParseAllGJSON10-4                     30000         52014 ns/op
BenchmarkParseAllUnMarshal10-4                 20000         75115 ns/op
BenchmarkParseAllGJSON100-4                     3000        415557 ns/op
BenchmarkParseAllUnMarshal100-4                 2000        589300 ns/op
BenchmarkParseAllGJSON1000-4                     300       3887135 ns/op
BenchmarkParseAllUnMarshal1000-4                 300       5813146 ns/op
BenchmarkParseAllGJSON5000-4                     100      20089087 ns/op
BenchmarkParseAllUnMarshal5000-4                  50      29568933 ns/op
BenchmarkParseAllGJSON10000-4                     30      40530862 ns/op
BenchmarkParseAllUnMarshal10000-4                 20      58910216 ns/op

```

**Comment by harshavardhana** [comment]
This issue is now closed by #2631
